### PR TITLE
Move fh-js-sdk to use upstream sync client

### PR DESCRIPTION
**Pull request by wtrocki**
### Motivation

Remove sync related source code from fh-js-sdk and use upstream sync client.
Upstream repository: https://github.com/feedhenry/fh-sync-js

### Changes map
Removing sync from library
https://github.com/feedhenry/fh-js-sdk/commit/b37954d168906176c47bb6d564fb1208a210f36d

Mounting sync to global namespace
https://github.com/feedhenry/fh-js-sdk/commit/1990255d766e2a4314ecc9b2b5f52b3340e24050

### Progress

- [x] Cleanup repo from sync
- [x] Add sync to build process (browserify)
- [x] Add custom handler for sync (use fh.cloud)
- [x] Documentation update
- [x] Release RC version of fh-sync-js to npm
- [x] Platform testing
- [x] Publishing package (RC) (fh-js-sdk@2.19.0-rc1)
- [x] Updates in all required dependencies
- [x] Notify group about this change
- [x] Verify any changes needs to be replicated to new repository
- [ ] Merge
- [ ] Publish version

Ticket: https://github.com/feedhenry/fh-sync-js/issues/2

### CODE REVIEW

**Comment by wtrocki** [comment]
@jasonmadigan - Do I have 🍏 green light to decouple client from product point of view?

@vchepeli - do you have time to help me with basic verification for sync? 
This is Raincatcher related action - we do not want to use js-sdk anymore in raincatcher code.

**Comment by david-martin** [comment]
@wtrocki I'm :+1:  on this if this means the $fh API is 100% backwards compatible  i.e. a build of this branch will work with existing apps using the previous version of the sdk.

**Comment by aidenkeating** [comment]
@wtrocki Taking a look now :)

**Comment by wtrocki** [comment]
@david-martin only one difference between is that sync will no longer use fh.act - just fh.cloud. Act was deprecated for couple years and after consultation with @wei-lee we decided that we no longer need it. Rest of the code is 100 % the same.

File: src/modules/sync_cloud_handler.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+var cloudAPI = require("./api_cloud");
+
+module.exports = function (params, success, failure) {
+    cloudAPI({
+        'path': '/mbaas/sync/' + params.dataset_id,
+        'method': 'post',
+        'data': params.req
+    }, function (res) {
+        callbackCalled = true;
```
Comment (wei-lee) [inline src/modules/sync_cloud_handler.js]: what's this for?

File: test/tests/test_sync_manage.js [hunk 2]
```diff
@@ -1,114 +0,0 @@
...
-        expect(config.sync_active).to.be.equal(true);
-        done();
-      }, function(message) {
-        expect(message).to.not.exist;
-        done(message);
-      }
-    );
-  });
-
-});
```
Comment (wei-lee) [inline test/tests/test_sync_manage.js]: are those tests removed to fh-sync-js now? I am wondering if we can keep a simple test here to make it easier to verify that sync is still backward compatible

Comment (wtrocki) [inline test/tests/test_sync_manage.js]: Yes. I moved maintenance for this tests to sync - they are executed as part of the PR build so nothing changes. Really good idea with simple test. Thanks

**Comment by aidenkeating** [comment]
@wtrocki What is the plan for the TypeScript declarations for sync, I see they have been moved to fh-sync-js, but they are duplicated here. That seems like it will cause issues with maintenance of the declaration files?

**Comment by wtrocki** [comment]
@aidenkeating Actually previous version I had had attempt to bring and integrate type definition file, but I had problems to do this effectively. We cannot just remove sync stuff from here as users will use fh-js-sdk as whole with sync being internal part of it. Tradeoff is to always maintain definitions in fh-sync-js and then update them on release.

**Comment by jasonmadigan** [comment]
@wtrocki I'm definitely on-board with doing this. Modularising our SDKs like this has been on the radar with a long, long time and I'm glad to see this tackled with Sync as a first step 👍

File: test/tests/test_sync_manage.js (hunk 2 above)
Comment (wtrocki) [inline test/tests/test_sync_manage.js]: Done!

**Comment by wtrocki** [comment]
### Verification steps

1) Use sync acceptance testing project:
https://github.com/feedhenry/sync-acceptance-testing

2) npm link this version fh-js-sdk 

3) Run tests

File: package.json [hunk 3]
```diff
@@ -62,10 +62,11 @@
   "license": "Copyright (c) 2014 FeedHenry Ltd, All Rights Reserved.",
   "gitHead": "251a1f88a39f59f2652552ae245893a833faee71",
   "dependencies": {
-    "type-of": "~2.0.1",
+    "fh-sync-js": "1.0.0",
     "loglevel": "~0.6.0",
```
Comment (austincunningham) [inline package.json]: version bump

**Comment by wtrocki** [comment]
RC version published for tests: fh-js-sdk@2.19.0-rc1

**Comment by wtrocki** [comment]
## Self verification results :] 

- [x] Verified using acceptance tests
- [x]  Verified locally using example template. 
- [x] Verified on feedhenry mbaas target. 
- [x] Verified on OpenShift.

screen shot 2017-06-13 at 6 10 44 pm

File: test/tests/test_sync.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    setTimeout(function(){
+      cb();
+    }, 600);
+  });
+}
+
+describe("test sync framework cloud handler", function(){
+  this.timeout(10000);
+  var header = { "Content-Type": "application/json" };
+  var xhr, requests;
```
Comment (wtrocki) [inline test/tests/test_sync.js]: Linting is supported, but we simply do not lint unit tests.

Comment (paolobueno) [inline test/tests/test_sync.js]: ah, why not?

**Comment by aidenkeating** [comment]
@wtrocki Has the fh-sync-js code been tested as a standalone module? It looks like the fhparams (cuid) etc. wouldn't be sent and sync would not work? I might be missing something though?

**Comment by wtrocki** [comment]
@aidenkeating - Going to supply this params in upstream version. This version is working fine as it's using fh.cloud calls, which are supplying this parameters.

**Comment by camilamacedo86** [comment]
The code shows fine, it is just don't respecting the eslint rules. However, I have some doubts about this change. What exactly in the functionalities it can affect? All information about how works the fh.cloud, fh.act and the fh.sync into the Client API docs[1] will still the same after this change?

[1] - https://access.redhat.com/documentation/en-us/red_hat_mobile_application_platform_hosted/3/html-single/client_api/#fh-act

*Observations* It is really important that all changes be tracked into the JIRA for the product owner has the visibility and for the support team knows that something was changed if came a case for example. It is to hard if we have more than on place to looking for something that can affect the functionality if the things are not all mapped in the same place into the same way.

**Comment by wtrocki** [comment]
@camilamacedo86 This change is 100 % compatibile. We just moved code to separate repository. Nothing else happened here.

**Comment by wtrocki** [comment]
Rebased on top of the master and retested. Any concerns before merging this?

**Comment by wtrocki** [comment]
Change was verified by Camila and Paolo. If no concerns I'm going to merge that tommorow.

File: src/modules/sync_cloud_handler.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
+var cloudAPI = require("./api_cloud");
+
+module.exports = function (params, success, failure) {
+    cloudAPI({
+        'path': '/mbaas/sync/' + params.dataset_id,
+        'method': 'post',
+        'data': params.req
+    }, function (res) {
```
Comment (evanshortiss) [inline src/modules/sync_cloud_handler.js]: Could just pass `success` directly here? Same for `failure`, e.g:

```js
cloudApi({
  'path': '/mbaas/sync/' + params.dataset_id,
  'method': 'post',
  'data': params.req
}, success, failure);
```

**Comment by evanshortiss** [comment]
Had similar questions to @aidenkeating but you've cleared them all! 👍

**Comment by evanshortiss** [comment]
Should we make this a major version bump to be safe? For example, perhaps the new upstream client could have breaking changes. It sounds like you tested and it's safe, but just a suggestion.

**Comment by wtrocki** [comment]
@evanshortiss - Major version bump should be used to mark release with some breaking changes. 
We just moved existing code so there is no breaking changes.

**Comment by aidenkeating** [comment]
@wtrocki If it has been verified by others then I am happy for it to be merged and published. If you need any extra verification then I'd be happy to.

**Pull request merged**